docs_nnx/why.rst

```diff
@@ -0,0 +1,416 @@
+* Automatic and efficient `PRNG management <https://flax.readthedocs.io/en/latest/glossary.html#term-RNG-sequences>`_.
+* `Variable Metadata <https://flax.readthedocs.io/en/latest/api_reference/flax.linen/spmd.html#flax.linen.with_partitioning>`_
+  for SPMD annotations, optimizer metadata, etc.
+
+One choice we made was to use functional (``compact``) semantics for NN programming via the lazy initialization of parameters,
+this made for concise  implementation code and aligned our API with Haiku. However, this also meant that the semantics of
+modules and variables in Flax were non-pythonic and often surprising. It also led to implementation complexity and obscured
+the core ideas of transformations on neural nets.
+
+.. testsetup:: Linen, NNX
+
+    import jax
+    from jax import random, numpy as jnp
+    from flax import nnx
+    import flax.linen as nn
+
+Introducing Flax NNX
+--------------------
+Flax NNX is an attempt to keep the features that made Linen useful while introducing some new principles.
+The central idea behind Flax NNX is to introduce reference semantics into JAX. These are its main features:
+
+- **Pythonic**: supports regular Python semantics for Modules, including for mutability and shared references.
+- **Simple**: many of the complex APIs in Flax Linen are either simplified using Python idioms or removed entirely.
+- **Better JAX integration**: both by making custom transforms adopt the same APIs as JAX transforms, and by making
+  it easier to use JAX transforms directly.
+
+Here's an example of a simple Flax NNX program that illustrates many of the points above:
+
+.. testcode:: NNX
+
+  from flax import nnx
+  import optax
+
+
+  class Model(nnx.Module):
+    def __init__(self, din, dmid, dout, rngs: nnx.Rngs):
+      self.linear = nnx.Linear(din, dmid, rngs=rngs)
+      self.bn = nnx.BatchNorm(dmid, rngs=rngs)
+      self.dropout = nnx.Dropout(0.2, rngs=rngs)
+      self.linear_out = nnx.Linear(dmid, dout, rngs=rngs)
+
+    def __call__(self, x):
+      x = nnx.relu(self.dropout(self.bn(self.linear(x))))
+      return self.linear_out(x)
+
+  model = Model(2, 64, 3, rngs=nnx.Rngs(0))  # eager initialization
+  optimizer = nnx.Optimizer(model, optax.adam(1e-3))  # reference sharing
+
+  @nnx.jit  # automatic state management for JAX transforms
+  def train_step(model, optimizer, x, y):
+    def loss_fn(model):
+      y_pred = model(x)  # call methods directly
+      return ((y_pred - y) ** 2).mean()
+
+    loss, grads = nnx.value_and_grad(loss_fn)(model)
+    optimizer.update(grads)  # in-place updates
+
+    return loss
+
+Improvements
+------------
+Through the rest of this document, we'll key examples of how Flax NNX improves on Flax Linen.
+
+Inspection
+^^^^^^^^^^
+The first improvement is that Flax NNX modules are regular Python objects, so you can easily
+construct and inspect them. Because Flax Linen Modules are lazy some attributes are not available
```

Comment:
nit:
> Because Flax Linen Modules are lazy some attributes are not available upon construction and are only accesible at runtime, this makes it hard to inspect and debug.

"Because Flax Linen Modules are lazy, some attributes are not available upon construction and are only accesible at runtime. This makes it hard to inspect and debug."